### DISCUSSION TITLE
Run Backup Task with Root Permission

### DISCUSSION THREAD

**Discussion by Wikinaut** [borgbase/vorta#801] [Ideas]
I usually run Vorta as a user. However for a complete backup of e.g. the ```/etc``` files, I think, it will be necessary to run Vorta as ```root```.

My wish and suggestion is that you describe the best way to perform such backups in your description or readme.

**Comment by Wikinaut** [comment]
For example, will a mere running as 
```
sudo /home/user/.local/bin/vorta
```
be sufficient to be able to backup all files including the files belonging to root?

**Comment by luckyrat** [comment]
I'd also appreciate some further guidance or links to relevant external documentation.

Ultimately I would like to see and use Vorta in the (non-root) user Gnome system tray (Ubuntu 18.04) with auto-start on system start (or at least Gnome login) and have the backups Vorta schedules ran with root permissions so that a full system backup can be taken.

To start with, I tried to just get it to run manually from a terminal window running in Gnome and failed:

````
<prompt>:~$ sudo vorta
sudo: vorta: command not found
<prompt>:~$ whereis vorta
vorta: /home/<user>/.local/bin/vorta
<prompt>:~$ sudo /home/<user>/.local/bin/vorta
Forking to background (see system tray).
<prompt>:~$ QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
````

At this point Vorta did not appear to be running but it may have been only the system tray part of it that was broken. I killed the background process and just ran as a normal user, which at least allows me to backup some files until I get this figured out...

````
<prompt>:~$ ps -ef | grep vorta
root      9612  4758  1 10:37 pts/0    00:00:00 /usr/bin/python3 /home/<user>/.local/bin/vorta
<user>  9641  8483  0 10:38 pts/0    00:00:00 grep --color=auto vorta
<prompt>:~$ sudo kill 9612
<prompt>:~$ vorta
Forking to background (see system tray).
<prompt>:~$ 
````

**Comment by ThomasWaldmann** [comment]
Just as an idea, did not test that yet:

if one uses the fat-binary of borg, it could be owned by root:root and be suid. Then vorta could run as normal user and borg would run as root.

This does NOT work with scripts though (e.g. shell or python), only with binaries.

**Comment by m3nu** [comment]
Not a fan of such workarounds. With Borg being suid, an attacker could change any file on the machine. (via `borg extract`)

Personally I'd either put your own user in some kind of "admin group" that has access to `/etc` or back up system files with Borgmatic running as root. The use case for Vorta is mostly to back up your personal files and home folder.

**Comment by luckyrat** [comment]
I guess that statements on the front page like "**One place** to view all point-in-time archives and restore individual files." gave me a different impression of what Vorta is for.

Maybe it's just me that interprets this as it being a complete backup solution for a system but if it is only supposed to be used for some files that's fine, I'll look into what this _Borgmatic_ alternative is instead. Thanks for the advice.

**Comment by m3nu** [comment]
Vorta is for desktops environments, not servers. Your use case belongs more to the server space. There, a headless solution, like Borgmatic or a shell script will work better with regards to permissions. 

If you're motivated you can surely run the backup task with `sudo`/`suid` and have full system access. It's just not a scenario we will support.

You can still use Vorta to access and restore those backup repos. I just don't recommend circumventing your local permissions and giving a user application deep system access without knowing your exact situation.

**Reply by NicholasFlamy** [reply]
> Vorta is for desktops environments, not servers. Your use case belongs more to the server space.

How is backing up a desktop a server use case? If I screw up my desktop/laptop in an update or I lose my laptop, I want to restore everything on a new machine or existing drive.

**Comment by Wikinaut** [comment]
@m3nu wrote
> Vorta is for desktops environments, not servers.

Äh, well, but.... for example some /etc files are still as important as user files. And databases.

So I think, as discussed above, at least the preferred-by-the-maintainers view should be explained. If you - for understandable reasons - do not suggest/recommend Vorta for these purposes, this should be mentioned.

**Comment by luckyrat** [comment]
> Vorta is for desktops environments, not servers. Your use case belongs more to the server space.

Fair enough. For context, I'm approaching the Linux backup in the same way as I have always backed up my Windows desktop environment (using tools that take regular differential backups of the entire system to enable both a quick bare-metal restore and fine-grained protection against short-term user file corruption/ransomware/human error). Given how hard I've found it to identify reliable software to achieve this I can entirely believe that it's just not a backup strategy that fits Linux as well as Windows and I will most likely learn more about better strategies as I read about Borgmatic.

I like what I've seen of Vorta so far and may well keep using it as part of a more complete backup solution.

**Comment by m3nu** [comment]
If you guys are really keen on doing backups as root, then I suggest to just run the `borg` command with `sudo`. So `borg create ...` becomes `sudo borg create`. Should be simple enough to configure passwordless sudo for just the Borg-command?

The PR for this should be quite simple. This option could be enabled under *Misc* or as env variable.

Would this solve your use case?

**Comment by luckyrat** [comment]
> Should be simple enough to configure passwordless sudo for just the Borg-command?

That's not something I've done before (sudo has always involved me typing a password).

I guess that this approach would still open up the risk of an attacker using `sudo borg extract` to overwrite arbitrary system files?

I think from a security perspective, the sort of solution I was dreaming of would involve a secure way (requesting root/sudo password for example) of configuring a script to run as root such that it can do a full backup of the system, with the scheduling and monitoring available via Vorta but without the ability for a user-account compromise to allow arbitrary root file access. The more I think about, the more it sounds like a fantasy world and I would think all similar systems I've used on Windows would have suffered from this same risk.

**Comment by Hofer-Julian** [comment]
> I think from a security perspective, the sort of solution I was dreaming of would involve a secure way (requesting root/sudo password for example) of configuring a script to run as root such that it can do a full backup of the system, with the scheduling and monitoring available via Vorta but without the ability for a user-account compromise to allow arbitrary root file access. The more I think about, the more it sounds like a fantasy world and I would think all similar systems I've used on Windows would have suffered from this same risk.

You can currently do that with a combination of [Timeshift](https://github.com/teejee2008/timeshift) and vorta.
I never used timeshift myself, but it is quite popular, several distros include it by default.

I don't know if it is feasible, to do something similar in vorta.

**Comment by m3nu** [comment]
> I guess that this approach would still open up the risk of an attacker using `sudo borg extract` to overwrite arbitrary system files?

No. The passwordless sudo can be restricted to [specific commands](https://unix.stackexchange.com/a/366420). So only `borg create` runs without password, but `borg extract` doesn't.

Only `borg create` would run with `sudo`. All other commands don't need it. (except for `borg check` maybe)

**Comment by Wikinaut** [comment]
(just let me drop it: thank you all - apart from working for/on borgbackup - for the - honestly - fruitful discussion of this topic.)

**Comment by ThomasWaldmann** [comment]
Well, "it depends".

## locally-mounted-filesystem repo

once you start doing "sudo borg create", your repo files will be owned by root and also your borg cache might be in /root and be owned by root (depends on sudo options). if you then use borg as non-root, you'll run into permission issues when accessing the repo files and either the same for the cache files or you'll have 2 different caches, one for root and one for the non-priv. user (might be a space and time consumption issue).

## ssh: repo

files in a ssh: remote repo are owned by the ssh login user, so no problem here.

one can use someuser@localhost to access local filesystem files and avoid the permission issues.

## borg extract

If one uses root to archive files not owned by the non-privileged user, one of course also needs root to correctly restore them (including file ownership and maybe also some other metadata). If you are a non-privileged user, you can't chown stuff away from yourself.

**Comment by m3nu** [comment]
Valid issues, @ThomasWaldmann . Didn't think of the cache and permissions.

The ssh-workaround mentioned already has its own issue #53. The generalized version would be "Set custom borg command", which could be `sudo borg` or `ssh root@localhost borg`.

To make this work, the user needs to set up SSH keys on their local machine and be very disciplined with any command he runs. I'm fine with adding this for power-users who need it in special cases and know what they are doing.

**Comment by adatum** [comment]
If Vorta cannot, and is not intended to, backup system files and other users' files, I think this should be mentioned in the website and documentation. 

I was considering using Vorta, but my goal is to create a complete system backup of a desktop and laptop. Based on this discussion, perhaps Vorta is not the right tool for the task.

**Comment by leggiero** [comment]
Hi guys. I'm a new Vorta user and arrived in this thread as had the same issue of not being able to backup some paths that needs super user permission.
Please, consider allowing custom borg commands where we can specify sudo, or just add the option to run as sudo only the `borg create` command called by Vorta.
Unfortunately, I'm temporarly using a command line based solution, but I'm waiting for this feature to use Vorta again. Will be following this thread.
Also, if you need beta testers, I'm here :-)

**Comment by poleguy** [comment]
I am working on a complete restore solution for a combined desktop/server use case. I don't like to focus on backup, because that's not the valuable part. :-)

I am trying out borgmatic as recommended in this thread, because I noticed vorta "silently failed" to back up some of my files because it's not running as root, and gives no visible warning if there are permission errors in the log. You have to click the logs and read through them to know there is trouble.

However, I like vorta particularly for the gui restore...borgmatic, not so much.

I don't distinguish much between 'server' and 'desktop' machines. I run 'server' processes on my desktop much like a server, and like many users I want backups of /etc so restoring my machine to working order is not a tremendous nightmare when something goes wrong. I'd also like to back up data that might belong to other user accounts (e.g. minecraft server account data, other user accounts, scripts sitting in root home directory, etc.).

Even if I do use borgmatic, I'd want a way to easily restore, so vorta is nice for that. Researching how to restore via command line is maybe not fun when you are under the stress of something bad having just happened. 

Seeing the gui icon is also a nice way to remember that backups are running and borgmatic doesn't do this.

So if I'm going to use vorta for restore, I'd like to use vorta for restoring borgmatic as well. But maybe I'm wrong. I'm not sure on the solution to this yet.

I don't think an admin group is enough if you need more than just /etc. Also I think that backup and restore should be at the same security level. Otherwise I think you are putting a hole in security: sudo borg create followed by borg restore might let a (compromised) user account read files they don't otherwise have access to. I think it's safer to keep it all as root or all as user to promote clear security thinking.

It would be nice to have a discussion of the use case intent in the getting started page, and a recommendation how to back up other user accounts, /etc, databases, and other data, so that vorta plays nice in this use case.

Also a screen shot showing how to check the log after your initial backup to check for permission trouble... and maybe an automatic warning?: 

https://vorta.borgbase.com/usage/local/

If I find a nice approach, I can write it up. For now I'm thinking vorta directly for my main account,  borgmatic as root for /etc, other users, databases, and run vorta as root for restore of the borgmatic stuff as needed. I'm open to ideas. Has anyone worked this out?

**Comment by witten** [comment]
Late to the party here. If I can make any changes to the way that borgmatic creates archives to improve the ability of Vorta to restore them, please let me know. I'm happy to look into that. However I imagine restoring with Vorta doesn't solve the root/non-root thing—you'd still need a way to restore privileged files. One of the work-arounds mentioned above may help with that.

**Comment by m3nu** [comment]
> I am trying out borgmatic as recommended in this thread, because I noticed vorta "silently failed" to back up some of my files because it's not running as root, and gives no visible warning if there are permission errors in the log. You have to click the logs and read through them to know there is trouble.

We regard backup runs as complete, even if some files were unaccessible, because it's very very common. E.g. on macOS many folders, like calendar or contacts storage are not accessible by default.

We could still add  note in the docs or show a note in Vorta. Though this may be confusing.

**Reply by NicholasFlamy** [reply]
> We regard backup runs as complete, even if some files were unaccessible, because it's very very common. E.g. on macOS many folders, like calendar or contacts storage are not accessible by default.

I'm sorry, but that's an awful way of doing things. "We know it can't backup some things so we just don't tell you ANYTIME something isn't backed up."
This way of doing things makes it impossible to trust it. Either let users know that wasn't backed up or put exclusions for MacOS.

**Reply by m3nu** [reply]
Responding to a 6-year old comment? The current state is that if some files are unreadable, Vorta will log that the backup was completed with errors. This is how most other tools do it as well.

Hope this helps to address your concerns.

**Reply by NicholasFlamy** [reply]
Thanks, I was reading this thread and saw this and was so appalled I had to respond. I haven't tried Vorta yet because I wanted to be sure what I'm about to do is supported and this discussion isn't closed (or doesn't appear to be) so I assumed that at least some of it is still relevant.

**Comment by amarendra** [comment]
> on macOS many folders, like calendar or contacts storage are not accessible by default

Indeed. 

While restic reads them and backs up just fine I have not been able to do it via borg (I run it from a script). I gave both borg (both original location - installed via Homebrew and symlinked path) and my script full disk access but it still can't read those paths.

**Comment by ThomasWaldmann** [comment]
@amarendra borg does nothing special, just open/read/close. If it can't do that, it does not have sufficient permissions. You can confirm that by running it as root.

**Comment by amarendra** [comment]
@ThomasWaldmann I will try it. Thanks.

Yes, it indeed seems to have permission issues.

> \~/Library/Mail: scandir: [Errno 1] Operation not permitted: '\~/Library/Mail'

Fails similarly for anything inside ~/Library.

I was just wondering that I installed `borg` as the user that owns that folder, I am logged in as that user, I have given `borg` full disk access in settings then what is that I am missing?

**Comment by ThomasWaldmann** [comment]
I don't use macOS myself, so I can't help you with debugging that.

**Comment by witten** [comment]
These [instructions for running with full disk access on macOS](https://projects.torsion.org/witten/borgmatic/issues/293) are for borgmatic, but the same general approach may work for directly running Borg as well.

**Comment by amarendra** [comment]
Thanks @witten. It didn't work for me. Besides I had just set that up for testing - this would have been quite a long-winded fix that would have needed to be kept updated all the time with increased dependencies. I will just skip those folders and find out some other way to backup those `~/Library` folders, or add them to my Restic backup and/or Tarsnap backup. Or wait for Vorta to become more stable (assuming that can do it).

**Comment by m3nu** [comment]
@amarendra, this is neither a Vorta nor a Borg issue. It's a security feature of macOS. If you enable "Full Disk Access" for the said app, you can copy any file, if not, some folders remain restricted. Often those (IMAP- or CalDAV-synced) folders shouldn't be backed up anyways. You may have picked the wrong app to allow depending on the way you installed Vorta. To be sure I tested it once again and everything works as expected:

https://files.qmax.us/vorta/vorta-full-disk-access-web.mp4

**Comment by amarendra** [comment]
Yes right, that's a Mac feature. When I mentioned waiting for Vorta to become "more stable" I was not pointing to this issue at all. I apologise if it came across that way. In fact I assumed Vorta could this do this. The above mentioned borgmatic workaround tries to do the it by bundling the script as an app and since Vorta is an app bundle (I mean I can see .app so again I assume) I thought it could do that.

> Often those (IMAP- or CalDAV-synced) folders shouldn't be backed up anyways

But why is that? e.g. I see my messages are in plain text in `~/Library/Messages/Archive` and I think it would be good to back keep a backup. That's how I did it till it broke after OS upgrade.

> https://files.qmax.us/vorta/vorta-full-disk-access-web.mp4

Thak you :) This is exactly how Tarsnap was able to back it up just now. I anyway plan to try Vorta again this weekend so I will do this.

**Comment by samuel-w** [comment]
For Linux, we can use `pkexec` to obtain root permissions for GUI. What is the equivalent for macOS if we need root?

Edit: Actually, root isn't needed for full disk access on macOS. Didn't fully read the comments.

**Comment by palto42** [comment]
@samuel-w The solution with `pkexec`works in principle, I tried it on Solus with [polkit file](https://gist.github.com/palto42/1064ab6d95240545c56afd9cebd44171) and extra `.desktop` file.

One observation I made is that it would be useful if the tray icon and the tray menu would change to indicate that this instance of Vorta runs as root. Otherwise it could be confusing and the user may forget that he started it as root.

![image](https://user-images.githubusercontent.com/12280188/100540193-14017b00-323c-11eb-815d-c70d364ee3f0.png)

One caveat I found when running Vorta as root is that it can't open web URLs in a browser, but for me this is no real issue.

**Comment by samuel-w** [comment]
Currently working on a solution that runs only the create as root. Its a lot more complex because enviroment variables don't transfer over and I have to fix the permissions after.

**Comment by palto42** [comment]
@samuel-w, not sure I understand the use case of restricting root to `borg create`. This would allow to create new backups, but not the restore anything from the backup or prune old archives etc.

**Comment by samuel-w** [comment]
Its possible if you create the backup as root and then fix the permissions with chown

**Comment by samuel-w** [comment]
I want to reduce the scope of root usage as much as possible to reduce security issues.

**Comment by palto42** [comment]
Although I understand the security concept of restricting root access to the minimum required, it doesn't very useful to me for this application to restrict it to only `borg create` which then requires some follow-up actions via root/sudo (e.g. changing permissions) to get things working.
But maybe it will get more clear as the development progresses, I'll be patient and wait...

**Comment by palto42** [comment]
@samuel-w I think we have 2 different use cases for running Vorta as root in mind.

1. Run Vorta as root in background (daemon) to automatically run system backups
2. Run Vorta as root interactively to configure system backups, browse the archive or restore data

I was thinking about 2nd use case, because I currently do my system backups via Borgmatic and mainly looked for a nice GUI to browse the archive. But the first use case looks interesting to use Vorta for unattended non-GUI backups of the whole system.

**Comment by Alexander-Shukaev** [comment]
What's the final consensus on this topic?  Should the main instance run under `root` user to perform complete system backup?  Or would it be enough to just run `sudo vorta --create ...`?  I am also looking for general recommendation to perform continuous self-scheduled system backups.

**Reply by m3nu** [reply]
I don't recommend running Vorta with `sudo`. Some desktop integrations may not work any more.

Best option is probably to run Borg with setuid or wrap the `borg` command in  script that runs `sudo borg`?

Both options may have security implications.

**Reply by adatum** [reply]
> I am also looking for general recommendation to perform continuous self-scheduled system backups.

Use a different tool. Vorta does not seem intended for this. Maybe see if `borgmatic` suits your needs: https://projects.torsion.org/witten/borgmatic

**Reply by m3nu** [reply]
Right. We focus on backups of your user files (documents, workspace, desktop, etc).

If you need to back up a server or headless machine, you're better off with Borgmatic or Borg directly.

**Reply by Wikinaut** [reply]
#note

Running `vorta` (or `borg`) as root – when it was installed via `pip` and „lives“ in `$HOME/.local/bin` :

```
# once: place a symbolic link to `vorta` in a suited path such as `/usr/bin` where root sees it
sudo ln -s $HOME/.local/bin/vorta /usr/bin/vorta

# run with -E to get the correct environment
sudo -E vorta
```

The same applies when you want to run `borg` as root (as mentioned by @m3nu above) – when it was installed via `pip` – mutatis mutandis.

**Reply by ThomasWaldmann** [reply]
> place a symbolic link to `vorta` in a suited path such as `/usr/bin` where root sees it

Custom stuff should be rather in `/usr/local/bin/`, but you need to check whether that is in the PATH.

If you have stuff in `$HOME/.local/bin` you can of course add that to PATH instead of symlinking elsewhere.

**Reply by Wikinaut** [reply]
Then (only putting /home/user/.local/bin into path for root), I nevertheless get (as root, when typing `borg`)
```
Traceback (most recent call last):
  File "/home/user/.local/bin/borg", line 5, in <module>
    from borg.archiver import main
ModuleNotFoundError: No module named 'borg'
```


However as sudo-user, I get 
```
sudo -E borg
sudo: borg: Befehl nicht gefunden
```
So, my two lines do work, but your idea appears not to be the full solution: putting $HOME/.local/bin into the PATH is not enough, you need at least a symlink, okay? I am only referring to work as root. As user, that's fine with $HOME/.local/bin .

**Comment by mYnDstrEAm** [comment]
I think this is a very important issue and the key advantage of [BackInTime](https://github.com/bit-team/backintime) over Vorta. There's no good reason why Vorta shouldn't also be able to do backups of root-owned files.

It's a crucial feature which would enable [system backup&restore](https://unix.stackexchange.com/q/656393/233262) (/ full system backups) which is a critical element of any efficient&robust IT system so I'm really surprised there isn't really any proper (GUI-including) solution for GNU/Linux and that isn't already built into Vorta.

It shouldn't be needed to run Vorta as root to get it working – afaik it would even be bad practice (to run any GUI as root) from a security standpoint (even though that's how BiT implemented it). One could simply add directories to sources and maybe check an option like "Backup root-owned files too". If that option is checked the user would need to enter the root password (only) once, maybe using pkexec, whenever the backup profile config was changed which would then set up/alter the cron job (or however the automatic backups are implemented).

**Comment by tracure1337** [comment]
Yes, a vorta systemd unit to run backups as root would be highly appreciated. I only backup my home but there are quite a few docker volumes owned by root because they run android etc.

Vorta is for desktops only is loosing as argument pretty quickly. I noticed that the yay wrapper for pacman also has many cache files owned by root in a normal users homedir.

I hope  for a solution one day :)

**Comment by jaysonwcs** [comment]
I don't get the "Vorta is for desktop" argument. It's obvious that some user files are saved on root folders or are owned by root.
I'd like at least to backup my VM configurations and some other /etc files, and they are all owned by root.

I'm currently using BackInTime only for this, but BackInTime uses Rsync with hardlinks, and because of that I lose space when storing the backup.

I would like to have a GUI to make Borg backups with root permissions, and Vorta is the only one I know so far.

Is there any way to make possible to Vorta run as root and perform the scheduled backups on root owned files?

**Reply by m3nu** [reply]
This thread already has a few workarounds. If we would support adding a custom Borg path (like `sudo borg...`, that may solve it too. Until then a one-line shell script would do the same. It will still mess up your cache files, use the wrong SSH keys among other things. So not keen on adding this at all.

**Comment by jaysonwcs** [comment]
The @palto42 suggestion (using the **polkit** file to call through **pkexec**, and adding a **desktop** file to **~/.config/autostart** folder) worked for me. The only problem is that it asks for the root password every time I start the computer (and not only when I open the GUI).

**I'm ok with this for now.**

**Comment by graugelb** [comment]
It seems to me that the subject here is closed down, so I do apologize to wake "already" up now:
My Subject: 
Vorta/Borg use for LMDE5 backup for a stand alone single user private PC storage to an external USB disk.
Usecase: total disk backup mainly to restore the whole system, but potentially also for single files  in case of a desaster.

Statues
I use for many years macrium reflect - also for Linux system like Mint, but is slow  - with now total trust as of many perfect experiences with restore.
Now I try  as backup Timeshift + Vorta, a combination already mentioned before.

To have one system only:
Is it possible to modify Vorta (what has made me already possible to use Borg) so that a whole disk can be stored with  ...clickycklicky onyl...?
Same procedure as for the home files.

I can imagine, that Borg is not interested to put work in a non commercial usecase. So I would accept any answer. But an simple answer. Now I am searching and searching and reading and not much understanding beside maybe...

**Reply by real-yfprojects** [reply]
I think *borg* isn't focussed on commercial users at all. Borg can backup a list of folders. So you can pass the mountpoint to borg and it will backup the whole drive. For accessing folders and files that have permissions denying access to the logged in user you will have to run borg as a root user / using sudo.

**Reply by prahal** [reply]
The Timeshift suggestion was based on ignorance. Timeshift especially on BTRFS only takes snapshots on the local disk, thus when you lose the drive you lose it all. Nothing to do with a system backup&rescue.
And borgmatic has no restore UI or even CLI AFAIK.
The only system backup and restore client for borg... is well, "borg" CLI.

Note that Vorta GUI refuses to add a directory for which you don't have the list permission thus you cannot add a folder as a user and have the backend only run as root on these directories.

**Reply by witten** [reply]
>  And borgmatic has no restore UI or even CLI AFAIK.

https://torsion.org/borgmatic/docs/how-to/extract-a-backup/

**Comment by Iceman248** [comment]
I'm new to this and am struggling to grasp where exactly we stand with using Vorta for backing up files owned as root. I don't need the whole system, just some files from within my /home and /etc, maybe /usr that are owned by root. I did manage to run it with "sudo vorta", but some things are a bit wonky like some have mentioned. I didn't understand the polkit file thing. I only backup locally.

**Reply by jaysonwcs** [reply]
You save the file that palto42 said on his comment (here https://github.com/borgbase/vorta/discussions/801#discussioncomment-369245) on `/usr/share/polkit-1/actions/`. Then, you need just to call Vorta on the terminal (or with `Alt+F2` on Gnome) with the following command: `pkexec vorta`.
It will ask for root password and will open a second instance of Vorta as root user.
If you want to autostart it, you can create a desktop file on the autostart folder of your distribution/desktop environment (on Gnome is _/home/**your user**/.config/autostart/_).

On this link (https://gist.github.com/palto42/1064ab6d95240545c56afd9cebd44171) there's the content of a polkit file. You just copy it and create the file on `/usr/share/polkit-1/actions/` with the name `com.borgbase.Vorta.policy` (as sudo). Then, you create the desktop file on _/home/**your user**/.config/autostart/_ folder (in case you're using Gnome environments) with the name `com.borgbase.Vorta_root.desktop` and copy the contents of the second file shown on that link.

**It will ask you for root password every time your desktop starts though.** It's a small inconvenience, but I'm ok with that because I really prefer Vorta over borgmatic.

If you're using another distro or desktop environment, you'll need to search online where you can create the desktop file, but every environment has its own folder to autostart apps.

**Reply by Eeems** [reply]
A little bit of a necropost, but if you are using a arch based system, I threw together a AUR package for this: https://aur.archlinux.org/packages/vorta-root

**Reply by eliliam** [reply]
> A little bit of a necropost, but if you are using a arch based system, I threw together a AUR package for this: https://aur.archlinux.org/packages/vorta-root

Do you install this in addition to vorta or does it replace it?

**Reply by Eeems** [reply]
> > A little bit of a necropost, but if you are using a arch based system, I threw together a AUR package for this: https://aur.archlinux.org/packages/vorta-root
> 
> Do you install this in addition to vorta or does it replace it?

A quick look at the PKGBUILD will answer that question for you: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vorta-root#n16

**Comment by Taxicletter** [comment]
I have question about what exactly is in /etc. When I searched, it would be configuration files. Aren't configuration files not (also) in the .config folder in the home folder?
What I want to know is: if you DON'T back up /etc, what exactly do I loose? I have a backup in case of a major crash. So I can easily restore a system to keep on working. Are the files in /etc necessary for that? Does it have my e-mail-credentials or wifi-password? Or is there stored if I have a panel on top or at the bottom of my screen?

The problem with the permissions for me, is not the files that are not backed up. The problem is more, that I each time get a "back up with warnings" and then I should look at the logs if it's only those etc-permissions, or if there is a more serious problem. And since I'm no programmer, a lot of what's in the logs is comprehensible for me.

**Reply by knezi** [reply]
`~/.config` is for your personal configuration, whereas `/etc` for system-wide. Well this is hard to generalise, but all your user config should be in your home (e.g. panel on top/bottom, password...).

As for `/etc`, I can think of wifi passwords (at least for NetworkManager), login manager (e.g. auto login), systemd enabled units... The question is even if you back it up will you be able to restore the files later? Simply copying the entire directory over to the new destination will give conflicts.

I personally, back up my entire `/` and then only specifically pick directories that I'm missing. Impossible to do if you don't know what they mean.

**Comment by linuxturtle** [comment]
I just recently started using borg to do remote backups of my homelab, and discovered Vorta, and was pretty excited to have a super slick/easy/fast way for me and my family to backup our desktops/laptops as well, then I ran into this problem...  So, to summarize the thread I've read so far (for the convenience of any that come after me):

_If you want to backup any files that aren't readable by the user you're running the Vorta GUI as, you should probably choose a different backup GUI/solution (BackInTime was mentioned).  There are workarounds in this thread involving various ways to run Vorta or borg as root, but they're all kludges creating various security and convenience issues, and the Vorta maintainer doesn't really think this is a problem worth fixing, or at least doesn't see an easy way to fix it._

Please correct me if that summary isn't accurate.

**Reply by linuxturtle** [reply]
OK, so after setting up BackInTime on my desktop, I think a followup is in order.  It's hardly a panacea either.  The way Backintime solves this problem is similar to one of the Vorta workarounds, it has a "run as root" .desktop file for the GUI.  This is less of a security problem with Backintime than it would be for Vorta, because Backintime's GUI is only used for setup and recovery, and doesn't need to be running all the time.  Note that if you're setting up the schedule, and you're trying to back up a machine that isn't powered on at predictable times (i.e. desktop/laptop), be sure to choose "anacron" as your scheduling method.  Otherwise, it's run by root's crontab, and if your machine doesn't happen to be powered on at the scheduled time, that backup will be missed.

**Reply by eliliam** [reply]
Yeah I'm pretty disappointed in the maintainer's stance on not supporting backing up the actual system, as that's truly what most people need, not just their home folder backed up. I prefer Vorta over BackInTime, but honestly it's a hard requirement to be able to back up all my configs and other non-user owned files, as those hold a lot of very important information for configuring your system.

Edit: Thanks @linuxturtle for the tip about using `anacron`, I never knew that was an option, very convenient!

**Reply by linuxturtle** [reply]
Hmm, my original post reads as a bit critical of @m3nu (the Vorta maintainer), and I didn't mean it to be, so maybe an apology is in order from me.  I see his reasons for not wanting to fix this, as it's a complicated problem without a tidy solution.  Vorta is supposed to be a simple and easy to deploy GUI front-end for borg, but in order to fix this issue, it would have to either be insecure (run as root 24x7, while simultaneously showing up in the unprivileged user's GUI), or get a lot more complicated (probably have some kind of privileged daemon running, which the front-end would connect to).  There aren't any simple or neat solutions AFAICT.

@eliliam, humorously, backintime doesn't actually use anacron when you select anacron, it just sets up an "every 15min" cron job in root's crontab to run itself, then it does an anacron-type thing and decides if it'll actually run, or just exit, depending on how long ago it last ran :laughing:

**Reply by m3nu** [reply]
No worries. If there is something to improve to make it work better with root, we can always do that. I'm not too experienced with Linux desktops, working on Linux servers mainly and mac desktop.

E.g. we could add a suitable .desktop file to the repo that works with root and document that.

**Reply by linuxturtle** [reply]
The run-as-root .desktop workaround is easy, and documented well in @palto42's [gist](https://gist.github.com/palto42/1064ab6d95240545c56afd9cebd44171), but I'm not sure it's a good idea to put in the repository/packaging for general usage for everyone to run (it's dangerous, and inadvisable to run a GUI app as root 24x7, with the GUI controllable by an unprivileged user, unless you really understand the ramifications of doing so).  Backintime does it, but their GUI only runs during setup, and when you need to recover files.  Periodic backups are done via a root cron job running a backend non-gui app, and it doesn't depend on the GUI for scheduling or running backups.  I don't think there's a clean way to do this without a major rewrite of Vorta, separating frontend/backend, and only running the backend as root when authenticated via sudo.  That would be a lot of effort, and I can totally see the POV that it's not worth it to make Vorta work as a full-system backup, when it wasn't conceived that way.

**Comment by greenmoss** [comment]
TL;DR: I offer another use case for Vorta with full-disk/root access: OS X users who want to back up all of their files, and are tired of fighting with OS X full disk access restrictions. This alone was a big selling point for me to try out Vorta.

More blah-blah from me:

I have been using a custom shell script to run Borg backup for years now.

It's "OK", with caveats:
* My installation of Borg requires Python, which is from homebrew, which periodically updates, thus loses full-disk access, thus invisibly starts breaking my backups
* I had to spend the time to write a shell script
* I had to finagle OS X to run the shell script
* OS X keeps mutating in weird ways that invisibly break my shell script - full disk access, discouraging/hiding cron, etc
* Editing and updating the shell script is fragile

Vorta is **so close** to what I need! I hope the root access thing can get a nice, elegant solution that fits with the GUI. Great work so far!

If you fine people decide against root access, I'll be sad, and go back to hacking my sad shell script.

**Comment by keysmashes** [comment]
Has the possibility of granting the `CAP_DAC_READ_SEARCH` ambient capability to Vorta been considered? This would allow it to read all files, but without the ability to write to anywhere other than the places it would normally be able to. (Though this will work only on Linux, not on macOS.)

For example, for those using NixOS, this seems like it should work:

```nix
{
  security.wrappers.vorta = {
    owner = "root";
    group = "root";
    capabilities = "cap_dac_read_search+eip";
    source = "${pkgs.vorta}/bin/vorta";
  };
}
```

**Reply by m3nu** [reply]
This would be great for per-OS documentations. I would consider it part of install instructions.

**Reply by Iruwen** [reply]
Obviously you wouldn't want to give regular users system read access either though, that's only slightly better than full write access.

/e: actually it shouldn't work anyway because both the vorta and borg commands are scripts, not binaries. It would have to be set on the python interpreter binary, which obviously is an even worse idea.

**Comment by VictorVow** [comment]
I wonder if [GPU Screen Recorder](https://flathub.org/apps/com.dec05eba.gpu_screen_recorder)'s approach would be applicable here, where it requires the user to install it as a system Flatpak rather than user Flatpak in order to access certain features.
